test_themes/tests/test_crawl.py

```diff
@@ -51,3 +147,101 @@ def test_02_homepage_tour_every_theme(self):
+                name = view.key.split('.')[1]
+                parent_name = view.inherit_id.key.split('.')[1]
+                self.assertEqual(name, parent_name, "Theme's new page template customization %r should use the same name as their parent %r" % (view.key, view.inherit_id.key))
+
+    def test_render_templates(self):
+        errors = []
+        view_ids = set()
+        websites_themes = self.env['website'].get_test_themes_websites()
+        for website in websites_themes:
+            with MockRequest(self.env, website=website):
+                views = self.env['ir.ui.view'].search([
+                    '|', '|',
+                    ('key', 'like', f'{website.theme_id.name}.s_'),
+                    ('key', 'like', f'{website.theme_id.name}.configurator'),
+                    ('key', 'like', f'{website.theme_id.name}.new_page'),
+                ])
+                view_ids.update(views.ids)
+                for view in views:
+                    try:
+                        self.env['ir.qweb']._render(view.id)
+                    except Exception:
+                        errors.append("View %s cannot be rendered" % view.key)
+        _logger.info("Tested %s views", len(view_ids))
+        self.assertGreater(len(view_ids), 1250, "Test should have encountered a lot of views")
+        self.assertFalse(errors, "No error should have been collected")
+
+    def test_render_applied_templates(self):
+        View = self.env['ir.ui.view']
+        errors = []
+        classes_inventory = set()
+        view_count = 0
+        def check(theme_name, website):
+            with MockRequest(self.env, website=website):
+                views = View.search([
+                    '|', '|',
+                    ('key', 'in', [
+                        'website.snippets',
+                        'website.new_page_template_groups',
+                    ]),
+                    ('key', 'like', escape_psql('website.configurator_')),
+                    ('key', 'like', escape_psql('website.new_page_template_sections_')),
+                ])
+                for view in views:
+                    try:
+                        html_text = self.env['ir.qweb']._render(view.id)
+                        if not html_text:
+                            continue
+                        html_tree = html.fromstring(html_text)
+                        blocks_el = html_tree.xpath("//*[@id='o_scroll']")
+                        if blocks_el:
+                            # Only look at blocks in website.snippets
+                            html_tree = blocks_el[0]
+                        for el in html_tree.xpath('//*[@class]'):
+                            classes = el.attrib['class'].split(' ')
+                            classes_inventory.update(classes)
+                            if len(classes) != len(set(classes)):
```

Comment:
```suggestion
                            if len(classes) != len(set(classes)):
```
```suggestion
                            if classes != set(classes):
```
Same?

Reply:
No:
```py
>>> [1,2,3] == set([1,2,3])
False
```
Wrapping the `set` back inside a `list` works - but then my guess is that `len` is cheaper.

Reply:
Indeed, glad I did not change that (test would have failed anyway)